src/goto-harness/memory_snapshot_harness_generator.cpp

```diff
@@ -0,0 +1,310 @@
+#include <util/string2int.h>
+#include <util/string_utils.h>
+#include <util/symbol_table.h>
+
+#include "goto_harness_generator_factory.h"
+
+void memory_snapshot_harness_generatort::handle_option(
+  const std::string &option,
+  const std::list<std::string> &values)
+{
+  if(option == "memory-snapshot")
+  {
+    memory_snapshot = require_exactly_one_value(option, values);
+  }
+  else if(option == "initial-location")
+  {
+    const std::string initial_location =
+      require_exactly_one_value(option, values);
+
+    std::vector<std::string> start;
+    split_string(initial_location, ':', start, true);
+
+    if(
+      start.size() == 0 || (start.size() >= 1 && start.front().empty()) ||
+      (start.size() == 2 && start.back().empty()) || start.size() > 2)
+    {
+      throw invalid_command_line_argument_exceptiont(
+        "invalid initial location specification", "--initial-location");
+    }
+
+    function = start.front();
+
+    if(start.size() == 2)
+    {
+      location_number = optionalt<unsigned>(safe_string2unsigned(start.back()));
+    }
+  }
+  else
+  {
+    throw invalid_command_line_argument_exceptiont(
+      "unrecognized option for memory snapshot harness generator",
+      "--" + option);
+  }
+}
+
+void memory_snapshot_harness_generatort::validate_options()
+{
+  if(memory_snapshot.empty())
+  {
+    throw invalid_command_line_argument_exceptiont(
+      "option --memory_snapshot is required",
+      "--harness-type initialise-from-memory-snapshot");
+  }
+
+  if(function.empty())
+  {
+    INVARIANT(
+      location_number.has_value(),
+      "when `function` is empty then the option --initial-location was not "
+      "given and thus `location_number` was not set");
+
+    throw invalid_command_line_argument_exceptiont(
+      "option --initial-location is required",
+      "--harness-type initialise-from-memory-snapshot");
+  }
+}
+
+void memory_snapshot_harness_generatort::add_init_section(
+  goto_modelt &goto_model) const
+{
+  goto_functionst &goto_functions = goto_model.goto_functions;
+  symbol_tablet &symbol_table = goto_model.symbol_table;
+
+  goto_functiont &goto_function = goto_functions.function_map[function];
+  const symbolt &function_symbol = symbol_table.lookup_ref(function);
+
+  goto_programt &goto_program = goto_function.body;
+
+  const symbolt &symbol = get_fresh_aux_symbol(
+    bool_typet(),
+    id2string(function),
+    "func_init_done",
+    source_locationt(),
+    function_symbol.mode,
+    symbol_table);
+
+  const symbol_exprt &var = symbol.symbol_expr();
+
+  // initialise var in __CPROVER_initialize if it is present
+  const auto f_it =
+    goto_functions.function_map.find(CPROVER_PREFIX "initialize");
+
+  if(f_it != goto_functions.function_map.end())
+  {
+    goto_programt &goto_program = f_it->second.body;
+
+    auto init_it =
+      goto_program.insert_before(std::prev(goto_program.instructions.end()));
+
+    init_it->make_assignment(code_assignt(var, false_exprt()));
+  }
```

Comment:
Now that since #4001 you can say
`goto_program.insert_before(...,  goto_programt::make_assignment(...));`
The `instructiont::make_assignment(...)` will eventually go away.

Reply:
Done.